src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/AssemblyBuilderImpl.cs

```diff
@@ -0,0 +1,136 @@
+        private ModuleBuilderImpl? _module;
+
+        internal AssemblyBuilderImpl(AssemblyName name, IEnumerable<CustomAttributeBuilder>? assemblyAttributes)
+        {
+            ArgumentNullException.ThrowIfNull(name);
+
+            _assemblyName = name;
+
+            _module = new ModuleBuilderImpl(ModuleBuilderImpl.ManifestModuleName, this);
+
+            if (assemblyAttributes != null)
+            {
+                foreach (CustomAttributeBuilder assemblyAttribute in assemblyAttributes)
+                {
+                    SetCustomAttribute(assemblyAttribute);
+                }
+            }
+        }
+
+        public static AssemblyBuilderImpl DefineDynamicAssembly(AssemblyName name)
+            => new AssemblyBuilderImpl(name, null);
+
+        public static AssemblyBuilderImpl DefineDynamicAssembly(
+            AssemblyName name,
+            IEnumerable<CustomAttributeBuilder>? assemblyAttributes)
+                => new AssemblyBuilderImpl(name, assemblyAttributes);
+
+        private static void WritePEImage(Stream peStream, MetadataBuilder metadataBuilder, BlobBuilder ilBuilder) // MethodDefinitionHandle entryPointHandle when we have main method.
+        {
+            // Create executable with the managed metadata from the specified MetadataBuilder.
+            var peHeaderBuilder = new PEHeaderBuilder(
+                imageCharacteristics: Characteristics.Dll // Start off with a simple DLL
+                );
+
+            var peBuilder = new ManagedPEBuilder(
+                peHeaderBuilder,
+                new MetadataRootBuilder(metadataBuilder),
+                ilBuilder,
+                flags: CorFlags.ILOnly,
+                deterministicIdProvider: content => new BlobContentId(Guid.NewGuid(), 0x04030201)); // Const ID, will reexamine as project progresses.
+
+            // Write executable into the specified stream.
+            var peBlob = new BlobBuilder();
+            BlobContentId contentId = peBuilder.Serialize(peBlob);
+            peBlob.WriteContentTo(peStream);
+        }
+
+        public void Save(Stream stream)
+        {
+            if (_previouslySaved) // Cannot save an assembly multiple times. This is consistent with Save() in .Net Framework.
+            {
+                throw new InvalidOperationException(SR.CannotSaveMultipleTimes);
+            }
+
+            ArgumentNullException.ThrowIfNull(stream);
+
+            if (_assemblyName == null || _assemblyName.Name == null)
+            {
+                throw new InvalidOperationException();
+            }
+
+            if (_module == null)
+            {
+                throw new InvalidOperationException(SR.AModuleRequired);
+            }
+
+            // Add assembly metadata
+            var metadata = new MetadataBuilder();
+            metadata.AddAssembly( // Metadata is added for the new assembly - Current design - metadata generated only when Save method is called.
```

Comment:
```suggestion
            metadata.AddAssembly(
```
This comment does not say anything useful. Is there alternative design that we want to switch to?